### PR TITLE
Add lineage tracking into Python SDK

### LINKED ISSUE #60: Add basic lineage logging for model, data, and metrics in notebook
/kind feature

**Describe the solution you'd like**
Improve the current backend/PyClient to have a lineage tracking for notebook. For the demo book we should have information which dataset generates the model and the metrics.

### PR DESCRIPTION
**Pull request by zhenghuiwang**
Fixes #60 #6 
The actual changes are on the Python SDK:
- Expose the `Execution` object to `log_input` or `log_output`
- Make `Run` a simple grouping concept.
- Update the [notebook example ](https://github.com/zhenghuiwang/metadata/blob/cec82de6a1d6f33d9258929195f80698910a0857/sdk/python/demo.ipynb)to reflect this change and add a section of lineage tracking.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/metadata/69)

### CODE REVIEW

**Comment by zhenghuiwang** [comment]
Ready for your review!

/assign @neuromage @gaoning777 

/cc @hongye-sun

File: sdk/python/kfmd/metadata.py [hunk 1]
```diff
@@ -114,22 +113,74 @@ def __init__(self, workspace=None, name=None, description=None):
       workspace {Worspace} -- Requried workspace object.
```
Comment (neuromage) [inline sdk/python/kfmd/metadata.py]: typo: Required. Here and below.

File: sdk/python/kfmd/metadata.py [hunk 2]
```diff
@@ -114,22 +113,74 @@ def __init__(self, workspace=None, name=None, description=None):
...
     input or output of this execution.
 
-    Returns a run object for logging.
+    Returns an execution object for logging.
     """
     # TODO(zhenghuiwang): check each field's type and whether set.
-    self.workspace = workspace
+    self.id = None
     self.name = name
+    if workspace == None:
```
Comment (neuromage) [inline sdk/python/kfmd/metadata.py]: `if workspace is None:`

File: sdk/python/kfmd/metadata.py [hunk 3]
```diff
@@ -114,22 +113,74 @@ def __init__(self, workspace=None, name=None, description=None):
...
+      raise ValueError("'workspace' must be set.")
+    self.workspace = workspace
+    self.run = run
     self.description = description
+    self.create_time = get_rfc3339_time()
+    response = self.workspace._client.create_execution(
```
Comment (neuromage) [inline sdk/python/kfmd/metadata.py]: Why is _client private then? Should it be made public instead?

File: sdk/python/kfmd/metadata.py [hunk 4]
```diff
@@ -114,22 +113,74 @@ def __init__(self, workspace=None, name=None, description=None):
...
+        parent=self.EXECUTION_TYPE_NAME,
+        body=self.serialization(),
+    )
+    self.id = response.execution.id
 
-  def log(self, artifact):
+  def serialization(self):
```
Comment (neuromage) [inline sdk/python/kfmd/metadata.py]: `serialized()` or `as_serialized_execution`or `as_mlmd_execution` or something along those lines. This returns a proto I assume?

File: sdk/python/kfmd/metadata.py [hunk 5]
```diff
@@ -114,22 +113,74 @@ def __init__(self, workspace=None, name=None, description=None):
...
+      execution.custom_properties[
+          WORKSPACE_PROPERTY_NAME] = openapi_client.MlMetadataValue(
+          string_value=self.workspace.name)
+    if self.run != None:
+      execution.custom_properties[
+          RUN_PROPERTY_NAME] = openapi_client.MlMetadataValue(
+          string_value=self.run.name)
+    return execution
+
+  def log_input(self, artifact):
```
Comment (neuromage) [inline sdk/python/kfmd/metadata.py]: `log_input_artifact` reads a little better. Same below.

**Pull request merged**